File: solutions/problem_2357_make_array_zero_by_subtracting_equal_amounts/solution.cpp
/*
 * Problem 2357: Make Array Zero by Subtracting Equal Amounts
 * ========================================================
 * Difficulty: Easy
 * Tags: Array, Hash Table, Greedy, Sorting, Heap (Priority Queue), Simulation
 * Pattern: Heap / Priority Queue
 *
 * Time Complexity:  O(n log n)
 * Space Complexity: O(n)
 */

#include <queue>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int minimumOperations(vector<int>& nums) {
        // Heap/Priority Queue - O(n log k) time
        priority_queue<int, vector<int>, greater<int>> pq;
        for (int val : nums) {
            pq.push(val);
            if ((int)pq.size() > nums)
                pq.pop();
        }
        return pq.empty() ? 0 : pq.top();
    }
};

File: solutions/problem_2357_make_array_zero_by_subtracting_equal_amounts/solution.py
"""
Problem 2357: Make Array Zero by Subtracting Equal Amounts
========================================================
Difficulty: Easy
Tags: Array, Hash Table, Greedy, Sorting, Heap (Priority Queue), Simulation
Pattern: Heap / Priority Queue

Time Complexity:  O(n log n)
Space Complexity: O(n)
"""

class Solution:
    def minimumOperations(self, nums: List[int]) -> int:
        # Heap/Priority Queue - O(n log k) time
        import heapq
        if not nums:
            return 0
        # Min heap (negate for max heap)
        heap = []
        for val in nums:
            heapq.heappush(heap, val)
            if len(heap) > (nums if isinstance(nums, int) else len(nums)):
                heapq.heappop(heap)
        return heap[0] if heap else 0
